### ISSUE TITLE
Cannot complete multiple requests to the same endpoint using http.request

### ISSUE THREAD

**Issue by mvattuone** [CondeNast/copilot-util#7]
@paulfryzel - I'm hoping you might have some insight into an issue I am experiencing currently.

My issue is with a plugin that depends on the HTTP request interface in `copilot-util` - if you'll look ~`[here]`~, you can see that we are using it in this plugin, which serves as a proxy, to make requests to an endpoint.

When I make calls to this endpoint in the application utilizing this plugin, it seems as though the first request completes, but the second stays is pending until 15 seconds pass, in which case a timeout error is returned. This seems like an issue within our application, rather than the direct endpoint (i.e. `some-slow-host`) since I am able to make subsequent calls to it without error. If I try to call the two routes directly in quick succession via two different browser tabs, the first returns data, and the second remains stuck until I close the initial request and try again.

Looking at [this line](https://github.com/CondeNast/copilot-util/blob/ceac9f69ff827bb6735d0d9609d318762df0318c/lib/http/request/index.js#L32) of code, in particular, and reading the docs for Node's [HTTP](https://nodejs.org/api/http.html#http_class_http_agent) module, I am wondering if it has something to do with the single socket connection "blocking" an incoming additional request to the same endpoint, since they are placed in the same queue. 

I am wondering if there is a way that we can utilize something like `keepAlive` to allow for a persistent connection that will allow for multiple requests. For now, I am using a fork of the plugin repo that uses `isomorphic-fetch` but it feels strange that I can't use the utils that we have which provide additional features. 

Let me know if anything is unclear - thanks!

cc @jsocol @RDFroeber

**Comment by paulfryzel** [comment]
@mvattuone Note this is an open source repository so I've removed references to internal projects/hosts. Try to generalize the problem domain where possible.

---

While official, the wording within the Node.js `http.Agent` documentation is a little confusing. Two requests to the same endpoint will not block each other unless there are no available sockets (in which case any new request would technically be blocked). Note that `maxSockets` defaults to `Infinity` unless overwritten elsewhere or memory/OS constrained.

What happens is that there is a queue of incoming requests to the agent for a particular host. That isn't a queue to make the request itself necessarily; rather it's a queue to request a socket be opened for the request to utilize. As the documentation notes, those open sockets are then used, on a one-request-per-socket-basis to initiate the HTTP request(s). Using keep-alive settings you can [sometimes] optimize this for quick, successive requests to reuse a recent socket (per host) and avoid the extra TCP/handshake steps (it will not be reused and/or block if not available; another new socket will be used instead assuming it can be opened).

If you watch the agent itself when making 'slow' or delayed requests you might see it's sockets look something like:

```js
...
agent = new http.Agent()
http.request({host:'my-slow-domain'}, onResponse) // request1
http.request({host:'my-slow-domain'}, onResponse) // request2
http.request({host:'my-slow-domain'}, onResponse) // request3
```
```js
console.log(agent.sockets)

// {'my-slow-domain': [
//   Socket {...},
//   Socket {...},
//   Socket {...}
// ]}
```

If `my-slow-domain` took ~10s to respond. ~10s you would see the `onResponse` callback invoked three times (note, not request1 at ~10s, request2 ~20s, request3 ~30s as they will be executed asynchronously).

If a timeout were reached before then, you'd simply see the request terminated and socket closed. Furthermore, if request2 and request3 decided to return successfully after ~1s, you'd see `onResponse` invoked twice at ~1s and then again after ~10s from request1. Individual request success/failure will have no effect on the others unless you were to destroy the entire agent in the process or link them in some other way (e.g. a callback from requestX terminates requestN).

You can test this yourself using the library directly:

```js
cu.http.request({hostname:'hostname-in-question', path:'/questionable/path'})
  .then(cu.http.response)
  .then(_ => console.log(_.toString()))
```
Run the above a few times in the REPL with varied paths/timeouts/etc. Unless limited in available sockets (or pushing extremely high throughput such to affect the available CPU/memory/fds needed to initiate the requests), you should see the approximate request completion times generally unaffected by the outcome of others, even with the same base domain.

It is my guess that something is blocking/restricting your returns at some other point/codepath or there is something here not being addressed (lack of available sockets etc). While libraries like `isomorphic-fetch` (via `node-fetch` in this case) add some additional layers on top of `http.request` to conform to that WhatWG standard. Any other bugs in `copilot-util` aside, it's the same general approach at the level you're describing here.

**Comment by paulfryzel** [comment]
I'd be happy to help debug this further with you (whether it's in this library or not) if you can provide more data or direct reproduction scenarios using this library directly.

**Comment by mvattuone** [comment]
Thank you for this very detailed explanation, @paulfryzel! 

I am able to reproduce this via curl and if I call `cu.http.request` directly in a REPL. So at least the issue is now isolated to the handler for the route.

As such, it's entirely possible that something within that handler is blocking, but I'm not entirely sure how to debug this in an efficient manner. Would definitely love some pointers on that front. While I was able to create and log an http agent in the REPL, I was unable to visualize incoming or current sockets by logging `agent.sockets`.

One thing I did notice is that TTFB is at 15s when looking at the network, but I figure that isn't so much a clue as a symptom, since we already know the request isn't completing.
